examples/batched-solver/batched-solver.cpp

```diff
@@ -108,46 +150,200 @@ int main(int argc, char *argv[])
-    auto bA = share(bmtx::create(exec, nbatch, A.get()));
-    auto bb = bvec::create(exec, nbatch, b.get());
-    auto bx = bvec::create(exec, nbatch, x.get());
-    const RealValueType reduction_factor{1e-7};
+    const real_type reduction_factor{1e-6};
 
     // Create the batch solver factory
-    auto solver_gen = rich::build()
-                          .with_max_iterations(500)
-                          .with_residual_tol(reduction_factor)
-                          .with_relaxation_factor(relax_factor)
-                          .on(exec);
+    auto solver_gen =
+        solver_type::build()
+            .with_max_iterations(500)
+            .with_residual_tol(reduction_factor)
+            .with_tolerance_type(gko::stop::batch::ToleranceType::relative)
+            .with_preconditioner(gko::preconditioner::batch::type::jacobi)
+            .on(exec);
+
+    // Create a logger to obtain the iteration counts and "implicit" residual
+    //  norms for every system after the solve.
+    std::shared_ptr<const gko::log::BatchConvergence<value_type>> logger =
+        gko::log::BatchConvergence<value_type>::create(exec);
 
     // Generate the batch solver from the batch matrix
-    auto solver = solver_gen->generate(bA);
+    auto solver = solver_gen->generate(A);
 
+    // add the logger to the solver
+    solver->add_logger(logger);
     // Solve the batch system
-    solver->apply(lend(bb), lend(bx));
+    solver->apply(lend(b), lend(x));
+    // This is not necessary, but one might want to remove the logger before
+    //  the next solve using the same solver object.
+    // solver->remove_logger(logger.get());
 
-    auto one = gko::batch_initialize<bvec>(nbatch, {1.0}, exec);
-    auto neg_one = gko::batch_initialize<bvec>(nbatch, {-1.0}, exec);
-    auto res = gko::batch_initialize<real_bvec>(nbatch, {0.0}, exec);
-    bA->apply(lend(one), lend(bx), lend(neg_one), lend(bb));
-    bb->compute_norm2(lend(res));
+    // Compute norms of RHS and final residual to check the result
+    auto b_norm =
+        gko::batch_initialize<real_vec_type>(num_systems, {0.0}, exec);
+    b->compute_norm2(lend(b_norm));
+    auto host_b_norm = real_vec_type::create(exec->get_master());
+    host_b_norm->copy_from(lend(b_norm));
+    // we need constants on the device
+    auto one = gko::batch_initialize<vec_type>(num_systems, {1.0}, exec);
+    auto neg_one = gko::batch_initialize<vec_type>(num_systems, {-1.0}, exec);
+    // allocate and compute the residual
+    auto res = vec_type::create(exec, batch_vec_size);
+    res->copy_from(lend(b));
+    A->apply(lend(one), lend(x), lend(neg_one), lend(res));
+    // allocate and compute residual norm on the device
+    auto res_norm =
+        gko::batch_initialize<real_vec_type>(num_systems, {0.0}, exec);
+    res->compute_norm2(lend(res_norm));
+    // transfer residual norm to the host
+    auto host_res_norm = real_vec_type::create(exec->get_master());
+    host_res_norm->copy_from(lend(res_norm));
```

Comment:
Can be simplified, because internally, it should be able to use `make_temporary_clone`
```suggestion
    auto res_norm =
        gko::batch_initialize<real_vec_type>(num_systems, {0.0}, exec->get_master());
    res->compute_norm2(lend(res_norm));
```

Reply:
That does not actually work because I'm printing it out, not `write`ing it. That's because I found it annoying to display the matrix market header before each line just for looking at the residual norms.

Reply:
does batch_initialize need?
I think it can be dense create and fill

Reply:
Never mind, I made a mistake while testing it out; it works.